xdsl/dialects/csl_/csl_stencil.py

```diff
@@ -0,0 +1,271 @@
+    ):
+        super().__init__(
+            operands=[op],
+            attributes={"offset": offset, "offset_mapping": offset_mapping},
+            result_types=[result_type],
+        )
+
+    def print(self, printer: Printer):
+        printer.print(" ")
+        printer.print_operand(self.op)
+        printer.print_op_attributes(
+            self.attributes,
+            reserved_attr_names={"offset", "offset_mapping"},
+            print_keyword=True,
+        )
+
+        mapping = self.offset_mapping
+        if mapping is None:
+            mapping = range(len(self.offset))
+        offset = list(self.offset)
+
+        printer.print("[")
+        index = 0
+        for i in range(len(self.offset)):
+            if i in mapping:
+                printer.print(offset[index])
+                index += 1
+            else:
+                printer.print("_")
+            if i != len(self.offset) - 1:
+                printer.print(", ")
+        printer.print("]")
+
+        printer.print_string(" : ")
+        printer.print_attribute(self.op.type)
+
+    @classmethod
+    def parse(cls, parser: Parser):
+        temp = parser.parse_operand()
+
+        index = 0
+        offset = list[int]()
+        offset_mapping = list[int]()
+        parser.parse_punctuation("[")
+        while True:
+            o = parser.parse_optional_integer()
+            if o is None:
+                parser.parse_characters("_")
+            else:
+                offset.append(o)
+                offset_mapping.append(index)
+            if parser.parse_optional_punctuation("]"):
+                break
+            parser.parse_punctuation(",")
+            index += 1
+
+        attrs = parser.parse_optional_attr_dict_with_keyword(
+            {"offset", "offset_mapping"}
+        )
+        attrs = attrs.data if attrs else dict[str, Attribute]()
+        attrs["offset"] = stencil.IndexAttr.get(*offset)
+        if offset_mapping:
+            attrs["offset_mapping"] = stencil.IndexAttr.get(*offset_mapping)
+        parser.parse_punctuation(":")
+        res_type = parser.parse_attribute()
+        if not isa(res_type, memref.MemRefType[Attribute]):
+            parser.raise_error("Expected return type to be a memref")
+        return cls.build(
+            operands=[temp], result_types=[res_type.element_type], attributes=attrs
+        )
+
+    def verify_(self) -> None:
+        assert isa(self.op.type, memref.MemRefType[Attribute])
+        assert self.result.type == self.op.type.get_element_type()
+
+        # As promised by HasAncestor(ApplyOp)
+        trait = cast(HasAncestor, AccessOp.get_trait(HasAncestor, (stencil.ApplyOp,)))
+        apply = trait.get_ancestor(self)
+        assert isinstance(apply, stencil.ApplyOp)
+
+        # TODO This should be handled by infra, having a way to verify things on ApplyOp
+        # **before** its children.
+        # cf https://github.com/xdslproject/xdsl/issues/1112
+        apply.verify_()
```

Comment:
I'm not sure if this will not result in the verification errors being "attached" to the wrong operation right now, i.e. in wrong IR the error message of the invalid apply will be printed on this access instead of the apply...

Reply:
This is copy-pasta from the stencil dialect.

This verifier needs the apply's verification to be done before, so this is the current solution.
MLIR has two different verification hooks for this kind of situations, we only have one :shrug: